File: Java/src/com/company/DynamicProgramming/LC121_maxProfit.java
package com.company.DynamicProgramming;

public class LC121_maxProfit {
    public int maxProfit(int[] p) {
        if(p.length == 0 || p.length == 1) return 0;
        int min = p[0];
        int[] dp = new int[p.length];
        dp[0] = 0;


        for(int i = 1; i < dp.length; i++){
            dp[i] = Math.max(dp[i-1], p[i] - min);
            if(p[i] < min) min = p[i];
        }

        return dp[p.length - 1];

    }
}

File: python/Array/LC121_maxprofit.py
class Solution:
    def maxProfit(self, prices: List[int]) -> int:
#        O(n^2) TLE
#         m = float('-inf')
#         for i in range(len(prices)-1):
#             for j in range(i+1, len(prices)):
#                 m = max(m, prices[j] - prices[i])
#         if m > 0:
#             return m
        
#         return 0
#       O(n)
        m = float('inf')
        res = 0
        for i in range(len(prices)):
            m = min(m, prices[i])
            res = max(res, prices[i] - m)
        return res